velox/connectors/hive/storage_adapters/hdfs/HdfsWriteFile.cpp

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/storage_adapters/hdfs/HdfsWriteFile.h"
+#include <hdfs/OutputStream.h>
+#include <hdfs/hdfs.h>
+
+namespace facebook::velox {
+
+HdfsWriteFile::HdfsWriteFile(hdfsFS hdfs, std::string_view path) {
+  // Pass 0 to the parameters bufferSize/replication/blockSize, means to use the
+  // default configured value. This is commented in the hdfs.h/hdfsOpenFile
+  // function.
+  new (this) HdfsWriteFile(hdfs, path, 0, 0, 0);
+}
+
+HdfsWriteFile::HdfsWriteFile(
+    hdfsFS hdfs,
+    std::string_view path,
+    int bufferSize,
+    short replication,
+    int blockSize)
+    : hdfsClient_(hdfs), filePath_(path) {
+  hdfsFile_ = hdfsOpenFile(
+      hdfsClient_,
+      filePath_.c_str(),
+      Hdfs::CreateFlag::Create | Hdfs::CreateFlag::Append,
+      bufferSize,
+      replication,
+      blockSize);
+  VELOX_CHECK_NOT_NULL(
+      hdfsFile_,
+      "Failed to open hdfs file: {}, with error: {}",
+      filePath_,
+      std::string(hdfsGetLastError()));
+}
+
+void HdfsWriteFile::close() {
+  int flag = hdfsCloseFile(hdfsClient_, hdfsFile_);
+  VELOX_CHECK_EQ(
+      flag,
+      0,
+      "Failed to close hdfs file, with error:{}",
+      std::string(hdfsGetLastError()));
+}
+
+void HdfsWriteFile::flush() {
+  int success = hdfsFlush(hdfsClient_, hdfsFile_);
+  VELOX_CHECK_EQ(
+      success, 0, "Hdfs flush error: {}", std::string(hdfsGetLastError()));
+}
+
+void HdfsWriteFile::append(std::string_view data) {
+  write(data, data.size());
+}
+
+void HdfsWriteFile::write(std::string_view data, int64_t dataSize) {
```

Comment:
I don't see this API in WriteFile interface. How do you plan to use this?

Reply:
this function was used in my unit test, to cover the error path when it write failed, I'll try to use other way to cover error path and remove this.

Reply:
Can we move this method to the test? We would not be adding test-only methods here.

Reply:
OK, I will try some other way to cover this error path.